### PR TITLE
Renew certificate fix

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: include/aos/iam/certhandler.hpp [hunk 1]
```diff
@@ -26,6 +24,55 @@ namespace certhandler {
  */
 constexpr auto cIAMCertModulesMaxCount = AOS_CONFIG_CERTHANDLER_MODULES_MAX_COUNT;
 
+/**
+ * Max number of certificate subscriptions:
+ *
+ * 1 per client & server + cIAMCertModulesMaxCount for message proxy
+ */
+constexpr auto cIAMCertSubsMaxCount = cIAMCertModulesMaxCount * 2 + cIAMCertModulesMaxCount;
```
Comment (al1img) [inline include/aos/iam/certhandler.hpp:32]: It should be private of CertHandler const. Also unclear why we need +cIAMCertModulesMaxCount

File: src/iam/certhandler/certhandler.cpp [hunk 2]
```diff
@@ -140,21 +138,45 @@ Error CertHandler::GetCertificate(
...
     }
 
+    certReceiver.UpdateCert(certInfo);
+
     return ErrorEnum::eNone;
 }
 
+Error CertHandler::UnsubscribeCertReceiver(CertReceiverItf& certReceiver)
+{
+    LockGuard lock(mMutex);
```
Comment (al1img) [inline src/iam/certhandler/certhandler.cpp:167]: LockGuard lock {mMutex};

File: include/aos/iam/certhandler.hpp [hunk 3]
```diff
@@ -26,6 +24,55 @@ namespace certhandler {
...
+ * Certificate receiver interface.
+ */
+class CertReceiverItf {
+public:
+    /**
+     * Updates certificate.
+     *
+     * @param info certificate info.
+     */
+    void UpdateCert(const CertInfo& info)
```
Comment (al1img) [inline include/aos/iam/certhandler.hpp:44]: It should not be part of interface. The client should have access only to required methods.

File: include/aos/iam/certhandler.hpp (hunk 1 above)
Comment (mykola-kobets-epam) [reply include/aos/iam/certhandler.hpp:32]: 1 per client & server + cIAMCertModulesMaxCount for message proxy

File: src/iam/certhandler/certhandler.cpp (hunk 2 above)
Comment (mykola-kobets-epam) [reply src/iam/certhandler/certhandler.cpp]: updated in a separate pr for all the files.

File: include/aos/iam/certhandler.hpp (hunk 3 above)
Comment (mykola-kobets-epam) [reply include/aos/iam/certhandler.hpp]: interface updated

File: src/iam/certhandler/certhandler.cpp [hunk 4]
```diff
@@ -202,6 +241,23 @@ CertModule* CertHandler::FindModule(const String& certType) const
...
 
-} // namespace certhandler
-} // namespace iam
-} // namespace aos
+Error CertHandler::UpdateCerts(CertModule& certModule)
+{
+    CertInfo certInfo;
+
+    for (auto& subscription : mCertReceiverSubscriptions) {
+        auto err = certModule.GetCertificate(Array<uint8_t>(), Array<uint8_t>(), certInfo);
```
Comment (mlohvynenko) [inline src/iam/certhandler/certhandler.cpp:249]: Didn't get why GetCertificate is called on each iteration, not before the loop

File: src/iam/certhandler/certhandler.cpp [hunk 5]
```diff
@@ -155,28 +153,64 @@ Error CertHandler::GetCertificate(
...
+          });
+
+    return err;
+}
+
+Error CertHandler::CreateSelfSignedCert(const String& certType, const String& password)
+{
+    LockGuard lock(mMutex);
+
+    LOG_DBG() << "Create self signed cert: type = " << certType;
```
Comment (mlohvynenko) [inline src/iam/certhandler/certhandler.cpp:201]: Remove spaces, so
`Create self signed cert: type='

File: src/iam/provisionmanager/provisionmanager.cpp [hunk 6]
```diff
@@ -145,4 +145,18 @@ Error ProvisionManager::GetCert(
     return AOS_ERROR_WRAP(mCertHandler->GetCertificate(certType, issuer, serial, resCert));
 }
 
+Error ProvisionManager::SubscribeCertReceiver(const String& certType, certhandler::CertReceiverItf& certReceiver)
+{
+    LOG_DBG() << "Subscribe cert receiver: type = " << certType;
```
Comment (mlohvynenko) [inline src/iam/provisionmanager/provisionmanager.cpp:150]: Remove spaces surrounding `=`

File: src/iam/certhandler/certhandler.cpp (hunk 5 above)
Comment (mykola-kobets-epam) [reply src/iam/certhandler/certhandler.cpp]: removed in a separate commit for the file

File: include/aos/iam/certhandler.hpp [hunk 7]
```diff
@@ -26,6 +24,24 @@ namespace certhandler {
  */
 constexpr auto cIAMCertModulesMaxCount = AOS_CONFIG_CERTHANDLER_MODULES_MAX_COUNT;
 
+/**
+ * Certificate receiver interface.
+ */
+class CertReceiverItf {
+public:
+    /**
+     * Process certificate updates.
```
Comment (al1img) [inline include/aos/iam/certhandler.hpp:33]: Processes

File: include/aos/iam/certhandler.hpp [hunk 8]
```diff
@@ -92,6 +108,23 @@ class CertHandlerItf {
...
         = 0;
 
+    /**
+     * Subscribes certificates receiver.
+     *
+     * @param certType certificate type.
+     * @param certReceiver certificate receiver.
+     * @returns Error.
+     */
+    virtual Error SubscribeCertReceiver(const String& certType, CertReceiverItf& certReceiver) = 0;
```
Comment (al1img) [inline include/aos/iam/certhandler.hpp:118]: SubscribeCertChanged?

File: include/aos/iam/certhandler.hpp [hunk 9]
```diff
@@ -221,16 +271,34 @@ class CertHandler : public CertHandlerItf, private NonCopyable {
     virtual ~CertHandler();
 
 private:
+    // (Two subscriptions for grpc client & server in iam/sm/cm/um) * modules count +
+    // (one subscription for grpc client in message proxy) * modules count
+    static constexpr auto cIAMCertSubsMaxCount = cIAMCertModulesMaxCount * 2 + cIAMCertModulesMaxCount;
```
Comment (al1img) [inline include/aos/iam/certhandler.hpp:276]: Looks like in general we need only one subscription per storage. So cIAMCertModulesMaxCount should be enough. Also, we can't have this option as const. Add dedicated define to config.hpp with default value equals to modules max count.

Comment (mykola-kobets-epam) [reply include/aos/iam/certhandler.hpp:276]: On the setup with main/secondary node  cIAMCertModulesMaxCount * 2 wasn't enough as I remember. And it looks like mp doesn't use IAMServer to get certificates, as well as IAMClient. But I guess, sm/cm/um clients - use. 
So I came up to the following formular: 
cIAMCertModulesMaxCount <for the servers> +  cIAMCertModulesMaxCount*cMaxNodesNum <for grpc clients> 

So we don't need additional configuration, the less configuration values we have - the better.

File: include/aos/iam/certhandler.hpp (hunk 8 above)
Comment (mykola-kobets-epam) [reply include/aos/iam/certhandler.hpp]: yes, it would be more uniform. changed

File: include/aos/iam/certhandler.hpp (hunk 7 above)
Comment (mykola-kobets-epam) [reply include/aos/iam/certhandler.hpp]: fixed

File: include/aos/iam/certhandler.hpp (hunk 9 above)
Comment (al1img) [reply include/aos/iam/certhandler.hpp]: I still do not understand why we need 3*cIAMCertModulesMaxCount subscriptions. Could you provide the example?

Please add dedicated configuration into config and set it to appropriate default value. Having config value doesn't mean configuring it all the time. But it make either to change this value in the feature.

Comment (mykola-kobets-epam) [reply include/aos/iam/certhandler.hpp]: one for sm/um/cm grpc server in communication manager + one for iam grpc server in iammanager 
one for grpc client in sm/um/iam manager
And I was expected that message proxy also requested certificates using iam public interface. But it looks like it's not.
So we only require 2*cert-modules-max-count.

Comment (al1img) [reply include/aos/iam/certhandler.hpp]: I see.  MP uses SM and IAM, client certificates and probably will use UM client certificates as well. We can't predict future client configuration. That's why config option is REQUIRED. And should be set to 3*AOS_CONFIG_CERTHANDLER_MODULES_MAX_COUNT.

Comment (mykola-kobets-epam) [reply include/aos/iam/certhandler.hpp]: > Please add dedicated configuration into config and set it to appropriate default value. 
Added in a separate commit

**Review by al1img — APPROVED** [review]

**Review by mlohvynenko — APPROVED** [review]

File: src/iam/certhandler/certhandler.cpp [hunk 10]
```diff
@@ -155,33 +153,69 @@ Error CertHandler::GetCertificate(
...
     auto* module = FindModule(certType);
     if (module == nullptr) {
         return AOS_ERROR_WRAP(ErrorEnum::eNotFound);
     }
 
-    auto err = module->CreateSelfSignedCert(password);
+    CertInfo certInfo;
+
+    auto err = module->GetCertificate(Array<uint8_t>(), Array<uint8_t>(), certInfo);
+    if (!err.IsNone()) {
```
Comment (mlohvynenko) [inline src/iam/certhandler/certhandler.cpp:170]: Consider this change
```suggestion
    if (!err.IsNone() && !err.Is(ErrorEnum::eNotFound)) {
```

**Review by MykolaSolyanko — APPROVED** [review]

File: src/iam/certhandler/certhandler.cpp (hunk 10 above)
Comment (mykola-kobets-epam) [reply src/iam/certhandler/certhandler.cpp:170]: agreed not to do that

**Pull request merged**